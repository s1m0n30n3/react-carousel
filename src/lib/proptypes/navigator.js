import PropTypes from "prop-types";

import { navigationPropTypes, galleryModeProps } from "./common";

export const navigatorPropTypes = {
  navigation: PropTypes.shape(navigationPropTypes),
  arrows: PropTypes.shape({
    direction: PropTypes.oneOf(["top", "right", "bottom", "left"]),
    icon: PropTypes.func,
    iconSize: PropTypes.number,
    color: PropTypes.string,
    background: PropTypes.string,
    size: PropTypes.number,
    radius: PropTypes.number,
  }),
  onNavigate: PropTypes.func,
  imagesLength: PropTypes.number,
  currentIndex: PropTypes.number,
  galleryMode: PropTypes.oneOf(galleryModeProps),
};
