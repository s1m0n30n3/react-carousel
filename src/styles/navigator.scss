.navigator--base {
    display: flex;
    justify-content: space-between;
    align-items: center;
    position: absolute;
    left: 0;
    right: 0;
    height: 0;
    margin: auto;
}