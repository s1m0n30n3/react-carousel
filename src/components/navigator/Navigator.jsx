import React from "react";

import { Arrow } from "components/arrow";
import { navigatorPropTypes } from "proptypes";
import { CarouselAction, GalleryMode } from "constants";
import { useNavPosition, useNavType } from "hooks";

import { isEmptyObject } from "helpers/isEmptyObject";

export const Navigator = ({
  navigation = {},
  arrows = {},
  onNavigate,
  imagesLength,
  currentIndex,
  galleryMode,
  ...props
}) => {
  const { type, position, width, sideMargin } = navigation;
  const { size, areSet, ...arrowProps } = arrows;

  const widthProp = `calc(${width} - ${sideMargin}px)`;

  const { style: navPosition } = useNavPosition({ size, position });
  const { style: navType } = useNavType({ width: widthProp, type });

  if (!areSet) return null;

  const isLastSlide = imagesLength - 1 === currentIndex;
  const isFirstSlide = currentIndex === 0;

  const mergedProps = { ...navType, ...navPosition };
  const style = isEmptyObject({ ...navType, ...navPosition })
    ? {}
    : { style: mergedProps };

  return (
    <div {...props} role="group" className="navigator--base" {...style}>
      <Arrow
        {...arrowProps}
        data-cy="arrow-left"
        size={size}
        disabled={galleryMode === GalleryMode.BOUNDED && isFirstSlide}
        direction="left"
        onClick={() => onNavigate({ action: CarouselAction.PREV })}
      />
      <Arrow
        {...arrowProps}
        data-cy="arrow-right"
        size={size}
        disabled={galleryMode === GalleryMode.BOUNDED && isLastSlide}
        direction="right"
        onClick={() => onNavigate({ action: CarouselAction.NEXT })}
      />
    </div>
  );
};

Navigator.propTypes = navigatorPropTypes;

Navigator.displayName = "Gallery.Navigator";
