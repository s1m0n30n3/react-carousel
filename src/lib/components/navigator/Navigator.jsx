import React from "react";

import { Arrow } from "components/arrow";
import { navigatorPropTypes } from "proptypes";
import { CarouselAction, GalleryMode } from "constants";
import { useNavPosition, useNavType } from "hooks";

import { base } from "styles/navigator.module.css";
import { isEmptyObject } from "helpers/isEmptyObject";

export const Navigator = ({
  navigationType: type,
  navigationPosition: position,
  navigationWidth: width,
  navigationSideMargin: margin,
  navigationArrows,
  onNavigate,
  imagesLength,
  currentIndex,
  galleryMode,
  ...props
}) => {
  const { arrowSize, areArrowsSet, ...arrowProps } = navigationArrows;
  const widthProp = `calc(${width} - ${margin}px)`;

  const { style: navPosition } = useNavPosition({ size: arrowSize, position });
  const { style: navType } = useNavType({ width: widthProp, type });

  if (!areArrowsSet) return null;

  const isLastSlide = imagesLength - 1 === currentIndex;
  const isFirstSlide = currentIndex === 0;

  const mergedProps = { ...navType, ...navPosition };
  const style = isEmptyObject({ ...navType, ...navPosition })
    ? {}
    : { style: mergedProps };

  return (
    <div {...props} role="group" className={base} {...style}>
      <Arrow
        {...arrowProps}
        data-cy="arrow-left"
        arrowSize={arrowSize}
        disabled={galleryMode === GalleryMode.BOUNDED && isFirstSlide}
        direction="left"
        onClick={() => onNavigate({ action: CarouselAction.PREV })}
      />
      <Arrow
        {...arrowProps}
        data-cy="arrow-right"
        arrowSize={arrowSize}
        disabled={galleryMode === GalleryMode.BOUNDED && isLastSlide}
        direction="right"
        onClick={() => onNavigate({ action: CarouselAction.NEXT })}
      />
    </div>
  );
};

Navigator.propTypes = navigatorPropTypes;

Navigator.displayName = "Gallery.Navigator";
