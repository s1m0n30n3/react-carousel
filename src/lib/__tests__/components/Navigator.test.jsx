import React from "react";
import renderer, { act } from "react-test-renderer";
import { describe, expect, test, vi } from "vitest";

import {
  NavigationType,
  NavigationPosition,
  GalleryMode,
  CarouselAction,
} from "constants";
import { Navigator } from "components/navigator";

import { ReactComponent as ArrowIcon } from "assets/icons/chevron-right.svg";

describe("Navigator", () => {
  test("does not render if `navigationArrows.areArrowSet` is equal to `false`", () => {
    const component = renderer.create(
      <Navigator
        navigationArrows={{
          areArrowsSet: false,
        }}
      />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("render if `navigationArrows.areArrowSet` is equal to `true` and `navigationArrows.arrowIcon` is defined", () => {
    const component = renderer.create(
      <Navigator
        navigationArrows={{
          areArrowsSet: true,
          arrowIcon: ArrowIcon,
        }}
      />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("render if `navigationArrows.areArrowSet` is equal to `true`, `navigationArrows.arrowIcon` and all other props are defined", () => {
    const component = renderer.create(
      <Navigator
        navigationType={NavigationType.ARROWS_AND_DOTS}
        navigationPosition={NavigationPosition.CENTER}
        navigationWidth="100%"
        navigationSideMargin={14}
        onNavigate={() => vi.fn(() => vi.mock({ action: CarouselAction.PREV }))}
        imagesLength={5}
        currentIndex={4}
        galleryMode={GalleryMode.RETURN}
        navigationArrows={{
          areArrowsSet: true,
          arrowSize: 12,
          arrowIcon: ArrowIcon,
          arrowIconSize: 24,
          arrowColor: "#ffffff",
          arrowBackground: "#000000",
          arrowRounding: 16,
        }}
      />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("render if `galleryMode` is equal to `GalleryMode.BOUNDED` and `imageLength - 1` === `currentIndex`, right button should be disabled", () => {
    const component = renderer.create(
      <Navigator
        imagesLength={5}
        currentIndex={4}
        galleryMode={GalleryMode.BOUNDED}
        navigationArrows={{
          areArrowsSet: true,
          arrowSize: 12,
          arrowIcon: ArrowIcon,
        }}
      />
    );

    const tree = component.toJSON();
    expect(tree.children[1].props.style.transform).toBe("rotate(0deg)");
    expect(tree.children[1].props.disabled).toBe(true);
    expect(tree.children[0].props.style.transform).toBe("rotate(180deg)");
    expect(tree.children[0].props.disabled).not.toBe(true);
    expect(tree).toMatchSnapshot();
  });

  test("render if `galleryMode` is equal to `GalleryMode.BOUNDED` and `currentIndex` === `0`, left button should be disabled", () => {
    const component = renderer.create(
      <Navigator
        imagesLength={5}
        currentIndex={0}
        galleryMode={GalleryMode.BOUNDED}
        navigationArrows={{
          areArrowsSet: true,
          arrowSize: 12,
          arrowIcon: ArrowIcon,
        }}
      />
    );

    const tree = component.toJSON();
    expect(tree.children[0].props.style.transform).toBe("rotate(180deg)");
    expect(tree.children[0].props.disabled).toBe(true);
    expect(tree.children[1].props.style.transform).toBe("rotate(0deg)");
    expect(tree.children[1].props.disabled).not.toBe(true);
    expect(tree).toMatchSnapshot();
  });

  test("render correctly and have the function `onClick` attached to first child `Arrow` component", () => {
    const component = renderer.create(
      <Navigator
        imagesLength={5}
        currentIndex={0}
        galleryMode={GalleryMode.BOUNDED}
        navigationArrows={{
          areArrowsSet: true,
          arrowSize: 12,
          arrowIcon: ArrowIcon,
        }}
        onNavigate={vi.fn()}
      />
    );

    const tree = component.toJSON();
    const { onNavigate } = component.toTree().props;
    const { onClick } = tree.children[0].props;

    act(() => onClick(() => onNavigate({ action: "prev" })));

    expect(onClick).toBeDefined();
    expect(onClick).toBeInstanceOf(Function);
    expect(onClick).toMatchSnapshot();
  });

  test("render correctly and have the function `onClick` attached to second child `Arrow` component", () => {
    const component = renderer.create(
      <Navigator
        imagesLength={5}
        currentIndex={0}
        galleryMode={GalleryMode.BOUNDED}
        navigationArrows={{
          areArrowsSet: true,
          arrowSize: 12,
          arrowIcon: ArrowIcon,
        }}
        onNavigate={vi.fn()}
      />
    );

    const tree = component.toJSON();
    const { onNavigate } = component.toTree().props;
    const { onClick } = tree.children[1].props;

    act(() => onClick(() => onNavigate({ action: "next" })));

    expect(onClick).toBeDefined();
    expect(onClick).toBeInstanceOf(Function);
    expect(onClick).toMatchSnapshot();
  });
});
