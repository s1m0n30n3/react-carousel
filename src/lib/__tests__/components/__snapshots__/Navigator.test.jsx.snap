// Vitest Snapshot v1

exports[`Navigator > does not render if \`navigationArrows.areArrowSet\` is equal to \`false\` 1`] = `null`;

exports[`Navigator > render correctly and have the function \`onClick\` attached to first child \`Arrow\` component 1`] = `[Function]`;

exports[`Navigator > render correctly and have the function \`onClick\` attached to second child \`Arrow\` component 1`] = `[Function]`;

exports[`Navigator > render if \`galleryMode\` is equal to \`GalleryMode.BOUNDED\` and \`currentIndex\` === \`0\`, left button should be disabled 1`] = `
<div
  role="group"
>
  <button
    data-cy="arrow-left"
    disabled={true}
    onClick={[Function]}
    style={
      {
        "height": "12px",
        "transform": "rotate(180deg)",
        "width": "12px",
      }
    }
  >
    <svg
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polygon
        points="7.293 4.707 14.586 12 7.293 19.293 8.707 20.707 17.414 12 8.707 3.293 7.293 4.707"
      />
    </svg>
  </button>
  <button
    data-cy="arrow-right"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "height": "12px",
        "transform": "rotate(0deg)",
        "width": "12px",
      }
    }
  >
    <svg
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polygon
        points="7.293 4.707 14.586 12 7.293 19.293 8.707 20.707 17.414 12 8.707 3.293 7.293 4.707"
      />
    </svg>
  </button>
</div>
`;

exports[`Navigator > render if \`galleryMode\` is equal to \`GalleryMode.BOUNDED\` and \`imageLength - 1\` === \`currentIndex\`, right button should be disabled 1`] = `
<div
  role="group"
>
  <button
    data-cy="arrow-left"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "height": "12px",
        "transform": "rotate(180deg)",
        "width": "12px",
      }
    }
  >
    <svg
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polygon
        points="7.293 4.707 14.586 12 7.293 19.293 8.707 20.707 17.414 12 8.707 3.293 7.293 4.707"
      />
    </svg>
  </button>
  <button
    data-cy="arrow-right"
    disabled={true}
    onClick={[Function]}
    style={
      {
        "height": "12px",
        "transform": "rotate(0deg)",
        "width": "12px",
      }
    }
  >
    <svg
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polygon
        points="7.293 4.707 14.586 12 7.293 19.293 8.707 20.707 17.414 12 8.707 3.293 7.293 4.707"
      />
    </svg>
  </button>
</div>
`;

exports[`Navigator > render if \`navigationArrows.areArrowSet\` is equal to \`true\` and \`navigationArrows.arrowIcon\` is defined 1`] = `
<div
  role="group"
>
  <button
    data-cy="arrow-left"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "transform": "rotate(180deg)",
      }
    }
  >
    <svg
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polygon
        points="7.293 4.707 14.586 12 7.293 19.293 8.707 20.707 17.414 12 8.707 3.293 7.293 4.707"
      />
    </svg>
  </button>
  <button
    data-cy="arrow-right"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "transform": "rotate(0deg)",
      }
    }
  >
    <svg
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polygon
        points="7.293 4.707 14.586 12 7.293 19.293 8.707 20.707 17.414 12 8.707 3.293 7.293 4.707"
      />
    </svg>
  </button>
</div>
`;

exports[`Navigator > render if \`navigationArrows.areArrowSet\` is equal to \`true\`, \`navigationArrows.arrowIcon\` and all other props are defined 1`] = `
<div
  role="group"
  style={
    {
      "bottom": 0,
      "top": 0,
      "width": "calc(100% - 14px)",
    }
  }
>
  <button
    data-cy="arrow-left"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "background": "#000000",
        "borderRadius": "16px",
        "fontSize": "24px",
        "height": "12px",
        "transform": "rotate(180deg)",
        "width": "12px",
      }
    }
  >
    <svg
      fill="#ffffff"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polygon
        points="7.293 4.707 14.586 12 7.293 19.293 8.707 20.707 17.414 12 8.707 3.293 7.293 4.707"
      />
    </svg>
  </button>
  <button
    data-cy="arrow-right"
    disabled={false}
    onClick={[Function]}
    style={
      {
        "background": "#000000",
        "borderRadius": "16px",
        "fontSize": "24px",
        "height": "12px",
        "transform": "rotate(0deg)",
        "width": "12px",
      }
    }
  >
    <svg
      fill="#ffffff"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polygon
        points="7.293 4.707 14.586 12 7.293 19.293 8.707 20.707 17.414 12 8.707 3.293 7.293 4.707"
      />
    </svg>
  </button>
</div>
`;
